<a href="https://colab.research.google.com/github/odumbuya1/deep-learning-challenge/blob/main/Copy_of_Solution_Optimization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##  1. Data Preprocessing (Please notes below)

In [1]:
# Import libraries and dependencies
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

In [2]:
#  Read charity_data.csv file to Pandas Dataframe
funded_Org_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
funded_Org_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the 'EIN' columns
funded_Org_df = funded_Org_df.drop(["EIN"],1)

<ipython-input-3-cea3a6c7198b>:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  funded_Org_df = funded_Org_df.drop(["EIN"],1)


In [4]:
# Determine the number of unique values for each column.
funded_Org_df.nunique()

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [5]:
# check the NAME value counts
name_counts = funded_Org_df['NAME'].value_counts()
#  check name counts greater than 5?
name_counts[name_counts>5]

PARENT BOOSTER USA INC                             1260
TOPS CLUB INC                                       765
UNITED STATES BOWLING CONGRESS INC                  700
WASHINGTON STATE UNIVERSITY                         492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC     408
                                                   ... 
OLD OAK CLIFF CONSERVATION LEAGUE INC                 6
AMERICAN NEPHROLOGY NURSES ASSOCIATION                6
HUMBLE ISD EDUCATIONAL SUPPORT GROUPS INC             6
PROFESSIONAL LOADMASTER ASSOCIATION                   6
CBMC INC                                              6
Name: NAME, Length: 354, dtype: int64

In [6]:
# Determine name counts less than or equal to 5, assign as names to replace.
names_to_replace = list(name_counts[name_counts <= 5].index)

# Names to replace in dataframe
for app in names_to_replace:
    funded_Org_df['NAME'] = funded_Org_df['NAME'].replace(app,"Other")

# Check if binning was successful
funded_Org_df['NAME'].value_counts()

Other                                           20043
PARENT BOOSTER USA INC                           1260
TOPS CLUB INC                                     765
UNITED STATES BOWLING CONGRESS INC                700
WASHINGTON STATE UNIVERSITY                       492
                                                ...  
HABITAT FOR HUMANITY INTERNATIONAL                  6
DAMAGE PREVENTION COUNCIL OF TEXAS                  6
FLEET RESERVE ASSOCIATION                           6
HUGH OBRIAN YOUTH LEADERSHIP                        6
INTERNATIONAL CONGRESS OF CHURCHES MINISTERS        6
Name: NAME, Length: 355, dtype: int64

In [7]:
# Determine APPLICATION_TYPE value counts for binning
application_type_counts = funded_Org_df['APPLICATION_TYPE'].value_counts()
application_type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [8]:
application_types_to_replace = list(application_type_counts[application_type_counts < 500].index)

# Review dataframe
for app in application_types_to_replace:
    funded_Org_df['APPLICATION_TYPE'] = funded_Org_df['APPLICATION_TYPE'].replace(app,"Other")

# Check if binning was successful
funded_Org_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [9]:
# View classification value counts for binning
class_count = funded_Org_df['CLASSIFICATION'].value_counts()
class_count

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [10]:
# Replace value if count is less than 1000
classes_to_replace = list(class_count[class_count < 1000].index)

# Review in dataframe
for cls in classes_to_replace:
    funded_Org_df['CLASSIFICATION'] = funded_Org_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
funded_Org_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [11]:
# Generate categorical variable lists
application_cat = funded_Org_df.dtypes[funded_Org_df.dtypes == "object"].index.tolist()

**Regenerate dataFrame with dummies**

In [12]:
funded_Org_with_dummies_df = pd.get_dummies(funded_Org_df)

In [13]:
# Accordingly assign preprocessed data into features (x) and target arrays (y) as follows:
X = funded_Org_with_dummies_df.drop(["IS_SUCCESSFUL"], axis='columns').values
y = funded_Org_with_dummies_df["IS_SUCCESSFUL"].values

# Use train_test_split method () to split preprocessed data into a training and testing datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [14]:
# Scale the training and testing features datasets by creating a
# StandardScaler instance, fitting it to the training data, then using
# the transform function.

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scaled dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

 **2: Create a Binary Classification Model: Compile, Train, and Evaluate the Model**

In [15]:
# Model declaration:- deep neural net
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  100
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10


# Assign the number of input features and nodes for each layer using TensorFlow and Keras.
# instantiate tensor flow
nn = tf.keras.models.Sequential()


# First hidden layer and activation function
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer and activation function
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="sigmoid"))

# Third hidden layer and activation function
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="sigmoid"))

# Output layer and activation function
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# model structure
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               39900     
                                                                 
 dense_1 (Dense)             (None, 30)                3030      
                                                                 
 dense_2 (Dense)             (None, 10)                310       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 43251 (168.95 KB)
Trainable params: 43251 (168.95 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [16]:
# Compile model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [ ]:
# Train model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

In [18]:
# Evaluate the model using the test data to determine the loss and accuracy.
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4462 - accuracy: 0.7887 - 512ms/epoch - 2ms/step
Loss: 0.44618749618530273, Accuracy: 0.788688063621521


In [19]:
# Save and export results to an HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


 4: **Report on the Neural Network Model**

---



 1. **Overview**

 The Neural Network model was built as a tool to identify applicants with prospects to succeed in their ventures using funding from Alphabet Soup Foundation.

 2. **Results**

 #### Data Preprocessing
----
**Q.1
What variable(s) are the target(s) for your model?**

Ans.1
The target variable for the model is (y) labelled "Is-Successful", which signifies if the funding / money was use effectively.
---
Q.2
What variable(s) are the features for your model? **bold text**

Ans.2
The features of the model are contained in the dataframe called (x) which includes the following columns: 'name', 'APPLICATION TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'INCOME_AMT','SPECIAL_CONSIDERATIONS', 'STATUS', and 'ASK_AMT'
---
Q3.
What variable(s) should be removed from the input data because they are neither targets nor features? **bold text**

Ans.3.:  EIN (Employer identificaiton) was dropped, to avoid ambiguity in matters of significance.
SPECIAL_CONSIDERATIONS was dropped - was not relevant.



#### Compiling, Training, and Evaluating the Model
----
Q.1
How many neurons, layers, and activation functions did you select for your neural network model, and why? **bold text**

1. Each of the three hidden layers had many neurons, which seeemed to improve accuracy. The number of epochs remained the same. The first activation function was 'relu' but the 2nd and 3rd were 'sigmoid'and the output function was 'sigmoid'. Changing the 2nd and 3rd activation functions to 'sigmoid' also helped boost the accuracy.
---

**Q2.
Were you able to achieve the target model performance?**

Ans.2
Yes, model achieved target performance.
---
Q3.
What steps did you take in your attempts to increase model performance? **bold text**

3. Changing name column and using sigmoids acticvation yielded improved efficiency gains

#### Summary
----

Just below the summary section here, see metrices used to evaluate model performance. These metrix include model_loss and model accuracy.

268/268: Indicates the number of batches processed out of the total number of batches during evaluation. In this case, it seems all 268 batches were processed

loss: 0.4486 indicates the calculated loss value, which is approximately 0.4486.
accuracy: 0.7879 indicates the accuracy achieved, which is approximately 78.79%.
Increased model accuracy facilitated correct classification using test data with an associated increased opportunity for successful applicants

526ms/epoch - 2ms/step:

526ms/epoch represents the time taken per epoch during evaluation.
2ms/step represents the time taken per step (or per batch) during evaluation.
Second Line Output:

The type of APPLICATION is one of the following; T3, T4, T5, T6, T7, T8, T10, and T19

The application has the following CLASSIFICATION; C1000, C2000, C3000, C1200, and C2100. The Random Foreste Classifier may be a good to consider inview of it high accuracy ratio of 0.776 as shown below.

Summary:
The model's evaluation output indicates a loss of approximately 0.4486 and an accuracy of approximately 78.79% on the test data.
These metrics are crucial in assessing how well the model performs on previously unseen data, providing insights into its generalization capability. Nevertheless, the Random Forest Model results shown below is recommended for deep data cleaning as it shows a performance level (using x_test_scale data) of 0.776 almost equivalent to the Neural Network model.





In [20]:
# Evaluate the model using the test data to determine the loss and accuracy.
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.4462 - accuracy: 0.7887 - 362ms/epoch - 1ms/step
Loss: 0.44618749618530273, Accuracy: 0.788688063621521


In [21]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

In [ ]:
# Create a random forest classifier.
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)

# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

# Evaluate the model
y_pred = rf_model.predict(X_test_scaled)
print(f" Random forest model accuracy: {accuracy_score(y_test,y_pred):.3f}")

 Random forest model accuracy: 0.776
